### ISSUE TITLE
Inference fails for sklearn's make_pipeline

### ISSUE THREAD

**Issue by bschnurr** [microsoft/pyright#1930]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

Pyright can't infer `model` when using "use Library Code For Types", even thought the call to `make_pipeline` is a one liner that just returns a class instance of type `Pipeline`

sklearn\pipeline.py 
```python
def make_pipeline(*steps, memory=None, verbose=False):
	return Pipeline(_name_estimators(steps), memory=memory, verbose=verbose)
```

Pylance doesn't have this issue because we added a special stub for 
the decorator _deprecate_positional_args
https://github.com/microsoft/pyrx/blob/00534fd7fc61c91fcff388a8ff38dd884493d3cc/packages/pylance-internal/bundled/stubs/sklearn-stubs/utils/validation.pyi#L36

I feel here we could add some short circuit logic to notice we dont need to figure out the parameters of TransformedTargetRegressor(..)  because we already know we `make_pipeline` is just returning a `Pipeline`

**To Reproduce**
Steps to reproduce the behavior.

```
pip install sklearn
```

```
import numpy as np
import scipy as sp
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import make_pipeline
from sklearn.compose import make_column_transformer
from sklearn.compose import TransformedTargetRegressor
from sklearn.linear_model import Ridge

categorical_columns = ['RACE', 'OCCUPATION', 'SECTOR',
                       'MARR', 'UNION', 'SEX', 'SOUTH']

preprocessor = make_column_transformer(
    (OneHotEncoder(drop='if_binary'), categorical_columns),
    remainder='passthrough'
)

model = make_pipeline(
    preprocessor,
    TransformedTargetRegressor(
        regressor=Ridge(alpha=1e-10),
        func=np.log10,
        inverse_func=sp.special.exp10
    )
)

```

**Expected behavior**

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
![image](https://user-images.githubusercontent.com/1946977/120243427-65ef8180-c21c-11eb-8c6b-909079589bac.png)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

**Additional context**

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. If I run pyright and install sklearn as indicated in the repro steps, pyright properly infers the type of `model` as `Pipeline`.

![Screen Shot 2021-05-31 at 4 40 50 PM](https://user-images.githubusercontent.com/7040122/120249201-f6cf5880-c22e-11eb-9646-45060bcde0a1.png)

**Comment by bschnurr** [comment]
hmm. might be windows only then. I'm seeing this on the benchmarker too.

requirements.txt
```
joblib==1.0.1
numpy==1.20.3
scikit-learn==0.24.2
scipy==1.6.3
sklearn==0.0
threadpoolctl==2.1.0
```

I see it on both python 3.9.2  and Python 3.7.7

**Comment by erictraut** [comment]
I wouldn't expect this to vary by platform. It's likely some other explanation.

Do you have `useLibraryCodeForTypes` enabled? It's off by default for pyright.

If you "Go To Declaration" for the symbol `make_pipeline`, where does it take you?

**Comment by bschnurr** [comment]
useLibraryCodeForTypes  enabled. 
goto def works

**Comment by bschnurr** [comment]
i'll try with the released version of pyright

**Comment by bschnurr** [comment]
```
[Info  - 8:46:39 AM] Pyright language server 1.1.145 starting
[Info  - 8:46:39 AM] Server root directory: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist
Waiting for Python extension to load
[Info  - 8:46:39 AM] Background analysis(1) root directory: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist
[Info  - 8:46:39 AM] Background analysis(1) started
Python extension loaded
Received pythonPath from Python extension: c:\Users\bschnurr\source\repos\pyright_sklearn\venv\Scripts\python.exe
[Info  - 8:46:46 AM] No configuration file found.
[Info  - 8:46:46 AM] No pyproject.toml file found.
[Info  - 8:46:46 AM] Setting pythonPath for service "pyright_sklearn": "c:\Users\bschnurr\source\repos\pyright_sklearn\venv\Scripts\python.exe"
[Warn  - 8:46:46 AM] stubPath c:\Users\bschnurr\source\repos\pyright_sklearn\typings is not a valid directory.
[Info  - 8:46:46 AM] Assuming Python version 3.7
[Info  - 8:46:46 AM] Assuming Python platform Windows
Search paths for c:\Users\bschnurr\source\repos\pyright_sklearn
  c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib
  c:\Users\bschnurr\source\repos\pyright_sklearn
  c:\Users\bschnurr\source\repos\pyright_sklearn\typings
  c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stubs\...
  c:\users\bschnurr\appdata\local\programs\python\python37\dlls
  c:\users\bschnurr\appdata\local\programs\python\python37\lib
  c:\users\bschnurr\appdata\local\programs\python\python37
  c:\users\bschnurr\source\repos\pyright_sklearn\venv
  c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages
[Info  - 8:46:46 AM] Searching for source files
[Info  - 8:46:46 AM] Auto-excluding c:\Users\bschnurr\source\repos\pyright_sklearn\venv
[Info  - 8:46:46 AM] Found 1 source file
[FG] parsing: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (39ms)
[FG] parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 4ms] (157ms)
[FG] binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\builtins.pyi (70ms)
[FG] binding: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (2ms)
Background analysis message: setConfigOptions
Background analysis message: ensurePartialStubPackages
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py ...
[BG(1)]   parsing: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (58ms)
[BG(1)]   parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 5ms] (205ms)
[BG(1)]   binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\builtins.pyi (59ms)
[BG(1)]   binding: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (1ms)
[BG(1)]   checking: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py ...
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\typing.pyi [fs read 2ms] (45ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\typing.pyi (31ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 0ms] (6ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (2ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 0ms] (5ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\typing_extensions.pyi (2ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\numpy\__init__.pyi [fs read 1ms] (69ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\numpy\__init__.pyi (25ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\scipy\__init__.py [fs read 1ms] (30ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\scipy\__init__.py (3ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\preprocessing\__init__.py [fs read 0ms] (4ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\preprocessing\__init__.py (1ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\preprocessing\_encoders.py [fs read 1ms] (57ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\preprocessing\_encoders.py (12ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\base.py [fs read 1ms] (56ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\base.py (5ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\pipeline.py [fs read 0ms] (19ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\pipeline.py (11ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\__init__.py [fs read 0ms] (1ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\__init__.py (0ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\_column_transformer.py [fs read 1ms] (29ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\_column_transformer.py (4ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\_target.py [fs read 1ms] (8ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\compose\_target.py (1ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\__init__.py [fs read 0ms] (2ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\__init__.py (0ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\_ridge.py [fs read 0ms] (52ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\_ridge.py (27ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\_base.py [fs read 0ms] (38ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\linear_model\_base.py (7ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\abc.pyi [fs read 0ms] (1ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\abc.pyi (0ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\utils\validation.py [fs read 0ms] (30ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\utils\validation.py (13ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\numpy\typing\__init__.py [fs read 1ms] (4ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\numpy\typing\__init__.py (5ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\functools.pyi [fs read 0ms] (10ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\functools.pyi (3ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\inspect.pyi [fs read 1ms] (10ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\inspect.pyi (7ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\warnings.pyi [fs read 0ms] (2ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\warnings.pyi (1ms)
[BG(1)]     parsing: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\utils\metaestimators.py [fs read 0ms] (6ms)
[BG(1)]     binding: c:\users\bschnurr\source\repos\pyright_sklearn\venv\lib\site-packages\sklearn\utils\metaestimators.py (2ms)
[BG(1)]     parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\collections\__init__.pyi [fs read 0ms] (13ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\collections\__init__.pyi ...
[BG(1)]       parsing: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\_collections_abc.pyi [fs read 0ms] (1ms)
[BG(1)]       binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\_collections_abc.pyi (1ms)
[BG(1)]     binding: c:\Users\bschnurr\.vscode\extensions\ms-pyright.pyright-1.1.145\dist\typeshed-fallback\stdlib\collections\__init__.pyi (10ms)
[BG(1)]   checking: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (906ms)
[BG(1)] analyzing: c:\Users\bschnurr\source\repos\pyright_sklearn\test.py (1234ms)
Background analysis message: resumeAnalysis

```

**Comment by erictraut** [comment]
I investigated this further. The symptoms previously led me to believe this was an issue with bidirectional type inference, but that was an incorrect diagnosis. The behavior is due to an optimization I added in the type checker's validation logic for call expressions. If it encounters errors when evaluating argument expressions, it avoids attempting to perform return type inference for the call. Return type inference can be very expensive, and if the input argument types are wrong to begin with, the inferred return type will likely be wrong as well. In this example, however, inference would have produced the correct result.

Another contributing factor in this case was an annotated decorator function applied to the constructor of the `TransformedTargetRegressor` class.

I've made two changes:
1. I've removed the optimization I mentioned above. Return type inference is always employed even in cases where argument expression evaluation results in errors. I'm curious to see whether that has a measurable performance impact.
2. I've enhanced the logic that detects decorator functions that are unannotated. When these are detected, the type checker will act as though the decorator is not present. Experience shows that this is better than attempting to infer the return type of the decorator, which often leads to confusing results.
